vsintegration/tests/FSharp.Editor.Tests/Refactors/RefactorTestFramework.fs

```diff
@@ -0,0 +1,99 @@
+﻿module FSharp.Editor.Tests.Refactors.RefactorTestFramework
+
+open System
+open System.Linq
+open System.Collections.Immutable
+open System.Collections.Generic
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Text
+open Microsoft.VisualStudio.FSharp.Editor
+open Microsoft.VisualStudio.FSharp.Editor.CancellableTasks
+
+open FSharp.Editor.Tests.Helpers
+open Microsoft.CodeAnalysis.CodeRefactorings
+open Microsoft.CodeAnalysis.CodeActions
+open System.Threading
+open FSharp.Compiler.Text
+
+let GetTaskResult (task: Tasks.Task<'T>) = task.GetAwaiter().GetResult()
+
+type TestCodeFix = { Message: string; FixedCode: string }
+
+type TestContext(Solution: Solution, CT) =
+    let mutable _solution = Solution
+    member this.CT = CT
+
+    member this.Solution
+        with set value = _solution <- value
+        and get () = _solution
+
+    interface IDisposable with
+        member this.Dispose() = Solution.Workspace.Dispose()
+
+    static member CreateWithCode(code: string) =
+        let solution = RoslynTestHelpers.CreateSolution(code)
+        let ct = CancellationToken false
+        new TestContext(solution, ct)
+
+    static member CreateWithCodeAndDependency (code: string) (codeForPreviousFile: string) =
```

Comment:
@psfinaki Small problem here. I get the following exception when debugging this : 
System.Exception: 'The file 'C:\test2.fs' was not part of the project. Did you call InvalidateConfiguration when the list of files in the project changed?'
But there is no such method on Project or Solution.

Reply:
Yeah so things get more complicated when we need more than one file for tests :(
Since it is all synthetic, things get out of sync between synthetic files, projects and solutions.

You can get some inspiration from these tests: https://github.com/dotnet/fsharp/blob/d16430f204da31492411cad4a5320a2b29429efa/vsintegration/tests/FSharp.Editor.Tests/FindReferencesTests.fs#L20

Let me know if you need any help. It should be doable, our FSharp.Test.ProjectGeneration framework is quite powerful in that regard, it is just not the best API there.

Reply:
Sooo.. there seems to be different behavior with multi file analyzing. 
AddExplicitReturnType.fs Line 51
```f#
let inferredType = memberFunc.ReturnParameter.Type.TypeDefinition.DisplayName
``` 

Im not happy with that deep access but  so far it worked.. Not with multi files. The TypeDefinition throws a InvalidOperationException "Not a named type". Seems memberFunc is too early in the compiler chain here ? 
The failing test actually creates the synthetic project on disk and when opening that, everything works.

Reply:
Hmm I would expect the above only depend on the type itself. As in, it would through for generic things like `'a`.

Reply:
@psfinaki Any Idea / approach or someone who could help? Yeah the compiler still treats it generically at this point, but we want the resolved type right? I looked at it a bit more but I dont know what the next step would be

Reply:
Hmm I'll recheck your branch tomorrow and try to hack a bit :)

Reply:
@SebastianAtWork I think I am getting there, just to clarify - in this context, what are the examples of code where you would like the refactoring to be applicable and where not?

Reply:
@SebastianAtWork soooo I remembered we are talking about the test `Correctly infer custom type that is declared earlier in project` and looked at that deeply - and I have some results. 

First good news is that the multi-file functionality does work currently:

https://github.com/dotnet/fsharp/assets/5451366/bcc01387-af5c-449e-9a48-f9f5e2143161

That means that the problem lies in the test. I spent a few hours debugging the crap out of it - now, I am _not entirely sure_ but I think the problem is related to the auto-generation of the module names.

IIRC, you have already discovered that you can actually see the generated synthetic project on the disc:
![image](https://github.com/dotnet/fsharp/assets/5451366/7f6253ce-948f-4070-9fb5-118903fa035c)

If you look at the files there, they have this module declaration auto-added, e.g.
```fsharp
module TestProject_134921ec.ModuleSecond

open ModuleFirst

let sum a b = {Value=a+b}
```

But then the refactoring and hence the compiler somehow deals with the code looking like this (that's what I see in the debug)
```fsharp
open ModuleFirst

let sum a b = {Value=a+b}
```

I think this kind of inconsistency explodes somewhere throwing this somewhat cryptic exception.

So what helped is the following modification to the test in question:
```fsharp
    let project =
        { SyntheticProject.Create(
            { sourceFile "First" [] with
                Source = myModule 
            },
            { sourceFile "Second" ["First"] with
                Source = code
            }) with

            AutoAddModules = false
        }
```
Once I modified the `expectedCode` to be:
```fsharp
module ModuleSecond

open ModuleFirst

let sum a b : MyType= {Value=a+b}
```
...the test started to pass without any other changes.

Reply:
Sure :) FYI maintainers can push to the forks but we rarely do so :D